### ISSUE TITLE
RealmSwift 95.0 Saving Lists Issue

### ISSUE THREAD

**Issue by austinllach** [realm/realm-cocoa#2533]
I ran into an issue with RealmSwift Lists.

I have three objects:

RealmItem
RealmItem2
RealmItem3

I have two goals:

First, my goal is to create a RealmItem that contains an array of RealmItem.

Second, my goal is to create a RealmItem3 that contains an array of RealmItem2.

Very simple code, but producing massive errors. Am I doing something wrong? Or can Realm not support this?

I appreciate any help. 

Here is code in test project to show errors. https://www.dropbox.com/s/arx62m12q9n609u/RealmSearch.zip?dl=0

```swift

import UIKit
import Foundation
import RealmSwift


class RealmItem: Object {
    dynamic var id = Int()
    dynamic var name = String()
    let group = List<RealmItem>()
    
    override static func primaryKey() -> String? {
        return "id"
    }
}

class RealmItem2: Object {
    
    dynamic var id = Int()
    dynamic var name = String()
    let group = List<RealmItem2>()
    
    override static func primaryKey() -> String? {
        return "id"
    }
    
}

class RealmItem3: Object {
    
    dynamic var id = Int()
    let group = List<RealmItem2>()
    
    override static func primaryKey() -> String? {
        return "id"
    }
}


class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        let item = RealmItem()
        item.name = "Bob"
        item.id = 1
        
        let item2 = RealmItem()
        item2.name = "Barry"
        item2.id = 2
        
        let item3 = RealmItem()
        item3.name = "Doug"
        item3.id = 3
        
        let group1 = RealmItem()
        group1.name = "First Group"
        group1.id = 500
        group1.group.append(item)
        group1.group.append(item2)
        group1.group.append(item3)
        
        let ex1 = RealmItem2()
        ex1.name = "One"
        ex1.id = 54
        
        let ex2 = RealmItem2()
        ex1.name = "Two"
        ex1.id = 243
        
        var array = [RealmItem2]()
        array.append(ex1)
        array.append(ex2)
        
        let group2 = RealmItem3()
        group2.id = 15
        
        
        for item in array{
            group2.group.append(item)
            
        }
        
        
        
        println("This is what Group 1 should look like")
        println(group1)
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        println("Group 2 doesn't even generate the first object")
        println(group2)
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        
        let realm = Realm()
        realm.write({
            
            realm.create(RealmItem.self, value: ["id": group1.id, "name": group1.name, "group": group1.group], update: true)
            realm.create(RealmItem3.self, value: ["id": group2.id, "exercises": group2.group], update: true)

        })
        

        
        println("Instead Group one saves 4 Objects")
        let group1Results = realm.objects(RealmItem)
        println(group1Results)
        println("End of Group 1")
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        
        
        println("Instead Group two is all messed up")
        let group2Results = realm.objects(RealmItem3)
        println(group2Results)
        println("End of Group 2")
        println("   ")
        println("   ")
        println("   ")
        println("   ")
        println("   ")

    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}
```

**Comment by austinllach** [comment]
Just figured out it has something to do with the primary keys... it works if I take primary keys off of the objects in the List...

don't know if this is a bug or just how it's suppose to work

**Comment by jpsim** [comment]
> Group 2 doesn't even generate the first object

Actually, what's happening here is that `ex1` is created and its properties set, then `ex2` is created, but its properties are never set, instead you're overriding `ex1`'s properties:

```swift
let ex1 = RealmItem2()
ex1.name = "One"
ex1.id = 54

let ex2 = RealmItem2()
ex1.name = "Two"
ex1.id = 243
```

So you end up with `ex1: { "Two", 243 }, ex2: { "", 0 }`. So that's expected.

> Instead Group one saves 4 Objects

Since the `RealmItem` model has a property of `List<RealmItem>`, you end up creating more "top-level" `RealmItem`s every time you add an entry to the `group` property. So you end up with something like this:

```
// realm.objects(RealmItem)
[
 { group1: [item, item2, item3] },
 { item },
 { item2 },
 { item3 }
]
// group1, item, item2 and item3 are all RealmItems, so they all show up when requesting "all RealmItem objects"
```

So that's expected as well.

> Instead Group two is all messed up

Your code prints this:

```
Results<RealmItem3> (
	[0] RealmItem3 {
		id = 15;
		group = RLMArray <0x7ff82376db00> (
		
		);
	}
)
```

Your write block wrote `RealmItem3` like this:

```swift
realm.create(RealmItem3.self, value: ["id": group2.id, "exercises": group2.group], update: true)
```

but the `RealmItem3` doesn't have an `exercises` property, so that dictionary entry is ignored from the input value. So all you're writing to the Realm is the `id`, which gets properly saved. This is all working as expecting as well.

So in conclusion, your code is running exactly as you wrote it, and Realm is behaving as expected.

Removing primary keys from your models has no impact on this behavior.

**Comment by austinllach** [comment]
wow, ha... thanks

I guess that's what happens when you stare at the same code too long. Much appreciated. Thanks again
